src/google/cloud/ndb/query.py

```diff
@@ -1618,6 +1615,369 @@ def _option(self, name, given, options=None):
+        deadline=None,
+        read_policy=None,  #  _datastore_api.EVENTUAL,  # placeholder
+        options=None,
+    ):
+        """Get the first query result, if any.
+
+        This is equivalent to calling ``q.fetch(1)`` and returning the first
+        result, if any.
+
+        Args:
+            keys_only (bool): Return keys instead of entities.
+            projection (list[str]): The fields to return as part of the query
+                results.
+            offset (int): Number of query results to skip.
+            limit (Optional[int]): Maximum number of query results to return.
+                If not specified, there is no limit.
+            batch_size (Optional[int]): Number of results to fetch in a single
+                RPC call. Affects efficiency of queries only. Larger batch
+                sizes use more memory but make fewer RPC calls.
+            prefetch_size (Optional[int]): Overrides batch size for first batch
+                returned.
+            produce_cursors (bool): Whether to generate cursors from query.
+            start_cursor: Starting point for search.
+            end_cursor: Endpoint point for search.
+            deadline (Optional[int]): Override the RPC deadline, in seconds.
+            read_policy: Defaults to `ndb.EVENTUAL` for potentially faster
+                query results without having to wait for Datastore to apply
+                pending changes to all returned records.
+            options (QueryOptions): DEPRECATED: An object containing options
+                values for some of these arguments.
+
+            Returns:
+                Optional[Union[entity.Entity, key.Key]]: A single result, or
+                    :data:`None` if there are no results.
+        """
+        raise NotImplementedError
+
+    def get_async(
+        self,
+        keys_only=None,
+        limit=None,
+        projection=None,
+        offset=None,
+        batch_size=None,
+        prefetch_size=None,
+        produce_cursors=False,
+        start_cursor=None,
+        end_cursor=None,
+        deadline=None,
+        read_policy=None,  #  _datastore_api.EVENTUAL,  # placeholder
+        options=None,
+    ):
+        """Get the first query result, if any.
+
+        This is the asynchronous version of :meth:`Query.get`.
+
+        Returns:
+            tasklets.Future: See :meth:`Query.get` for eventual result.
+        """
+        raise NotImplementedError
+
+    def count(
+        self,
+        keys_only=None,
+        limit=None,
+        projection=None,
+        offset=None,
+        batch_size=None,
+        prefetch_size=None,
+        produce_cursors=False,
+        start_cursor=None,
+        end_cursor=None,
+        deadline=None,
+        read_policy=None,  #  _datastore_api.EVENTUAL,  # placeholder
+        options=None,
+    ):
+        """Count the number of query results, up to a limit.
+
+        This returns the same result as ``len(q.fetch(limit))`` but more
```

Comment:
This is probably overselling any potential performance advantage. Datastore does not have any `count` API we can call, so we basically do have to run the query and just count the results passed back. The only optimizations I can see us being able to do here are we can set `keys_only` to always be true, and we can save on the step to marshall entities (or keys) from the raw protocol buffers. Generally, network is going be our bottle neck, though. I would be surprised if NDB were CPU bound.